### ISSUE TITLE
Training keypoint model

### ISSUE THREAD

**Issue by eshafeeqe** [NikolasEnt/soccernet-calibration-sportlight#6]
Hello, 

Thanks for open sourcing this project. I run the training code for the keypoints model and got following log

```[2024-02-21 00:44:21,848][INFO]: train - epoch: 0, lr: 0.0001, train_loss: 0.07756424
[2024-02-21 00:45:33,460][INFO]: val - epoch: 0, val_loss: 0.08277992, val_precision: 0, val_recall: 0, val_pcks-2.0: 0, val_pcks-5.0: 0, val_pcks-10.0: 0, val_pcks-50.0: 0, val_l2: inf, val_l2_reprojection: inf, val_completeness: 0, val_eval_precision: 0, val_eval_recall: 0, val_eval_accuracy: 0, val_evalai: 0
[2024-02-21 00:45:37,497][INFO]: Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/model-000-0.082780.pth'
[2024-02-21 00:45:40,893][INFO]: Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/evalai-000-0.000000.pth'
[2024-02-21 00:45:44,029][INFO]: Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/pcks-5.0-000-0.000000.pth'
[2024-02-21 01:12:53,469][INFO]: train - epoch: 1, lr: 0.0001, train_loss: 0.07734418
[2024-02-21 02:53:39,295][INFO]: val - epoch: 1, val_loss: 0.08279017, val_precision: 0, val_recall: 0, val_pcks-2.0: 0, val_pcks-5.0: 0, val_pcks-10.0: 0, val_pcks-50.0: 0, val_l2: inf, val_l2_reprojection: inf, val_completeness: 0, val_eval_precision: 0, val_eval_recall: 0, val_eval_accuracy: 0, val_evalai: 0
[2024-02-21 02:53:43,217][INFO]: Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/save-001.pth'
```
You can see that the training got stopped after one epoch. Is that normal ? Got validation loss of 0.08277, whether this numbers are good enough ?

**Comment by NikolasEnt** [comment]
Hi @eshafeeqe . Thank you for your interest in the project.

It is not entirely clear from the logs you've shared whether the training process has stopped. As for me, it looks like the training is progressing in the background. In general, more or less usable results are expected after ~50 epochs, so the first completed epoch you've shared is just the very beginning of training.
Overall, the number of epochs as well as early stopping policy are defined in of the [training config](https://github.com/NikolasEnt/soccernet-calibration-sportlight/blob/master/src/models/hrnet/train_config.yaml).

**Comment by eshafeeqe** [comment]
Thanks for the reply. 

I have one more query. I am running your training code and got following training log

```
[2024-03-13 07:17:54,626][argus.model.build_140640061333632][INFO] - val - epoch: -1, val_loss: 0.09949375, val_precision: 0, val_recall: 0, val_pcks-2.0: 0, val_pcks-5.0: 0, val_pcks-10.0: 0, val_pcks-50.0: 0, val_l2: inf, val_l2_reprojection: inf, val_completeness: 0, val_eval_precision: 0, val_eval_recall: 0, val_eval_accuracy: 0, val_evalai: 0
[2024-03-13 07:56:33,792][argus.model.build_140640061333632][INFO] - train - epoch: 0, lr: 0.0001, train_loss: 0.07753553
[2024-03-13 07:59:26,642][argus.model.build_140640061333632][INFO] - val - epoch: 0, val_loss: 0.08273235, val_precision: 0.999844, val_recall: 0.9959607, val_pcks-2.0: 0.2039847, val_pcks-5.0: 0.3618192, val_pcks-10.0: 0.3986444, val_pcks-50.0: 0.409502, val_l2: 6.074544, val_l2_reprojection: 13.92419, val_completeness: 0.491594, val_eval_precision: 0.680844, val_eval_recall: 0.9070278, val_eval_accuracy: 0.6913531, val_evalai: 0.3398651
[2024-03-13 07:59:29,132][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/model-000-0.082732.pth'
[2024-03-13 07:59:30,938][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/evalai-000-0.339865.pth'
[2024-03-13 07:59:33,182][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/pcks-5.0-000-0.361819.pth'
[2024-03-13 08:38:11,994][argus.model.build_140640061333632][INFO] - train - epoch: 1, lr: 0.0001, train_loss: 0.07742175
[2024-03-13 08:43:33,228][argus.model.build_140640061333632][INFO] - val - epoch: 1, val_loss: 0.08272079, val_precision: 0.9998931, val_recall: 0.9986119, val_pcks-2.0: 0.3038401, val_pcks-5.0: 0.5161353, val_pcks-10.0: 0.5769206, val_pcks-50.0: 0.5986358, val_l2: 5.808748, val_l2_reprojection: 7.75617, val_completeness: 0.6304483, val_eval_precision: 0.7497951, val_eval_recall: 0.9723649, val_eval_accuracy: 0.743344, val_evalai: 0.46864
[2024-03-13 08:43:36,026][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/model-001-0.082721.pth'
[2024-03-13 08:43:36,158][argus.model.build_140640061333632][INFO] - Model removed '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/model-000-0.082732.pth'
[2024-03-13 08:43:37,851][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/evalai-001-0.468640.pth'
[2024-03-13 08:43:37,995][argus.model.build_140640061333632][INFO] - Model removed '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/evalai-000-0.339865.pth'
[2024-03-13 08:43:39,979][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/pcks-5.0-001-0.516135.pth'
[2024-03-13 08:43:40,160][argus.model.build_140640061333632][INFO] - Model removed '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/pcks-5.0-000-0.361819.pth'
[2024-03-13 08:43:43,022][argus.model.build_140640061333632][INFO] - Model saved to '/workdir/data/experiments/HRNet_57_hrnet48x2_57_003/save-001.pth'
[2024-03-13 09:22:26,569][argus.model.build_140640061333632][INFO] - train - epoch: 2, lr: 0.0001, train_loss: 0.07730475
```
We can see that it taking around 40 minutes for one epoch in my system. Just wondering whether this is normal ? I am using rtx 3090 with 24gb of memory.

**Comment by eshafeeqe** [comment]
How you recommend to set the parameters of training_config ? Whether to run the training with parameters defined in the code and do hyper parameter optimisation after around 30 epochs  ?

**Comment by NikolasEnt** [comment]
Hi @eshafeeqe. The training time for the GPU seems to be reasonable, for me it was about 34 min per epoch on RTX 3090, but it may depend on various factors, like the GPU model itself (whether it has factory OC), disk and RAM speed, the CPU performance, etc.

The current set of params in [training config](https://github.com/NikolasEnt/soccernet-calibration-sportlight/blob/master/src/models/hrnet/train_config.yaml) allows end-to-end model training for the best achieved results we had during the competition, so, no optimization required for results reproduction. However, you may try to optimize it further.

**Comment by eshafeeqe** [comment]
Thanks for your help.
